### PR TITLE
Verify @example docs on npm test

### LINKED ISSUE #598: Is there any way to make our example code in docstrings part of the test suite?
I'm worried about code rot and typos -- is there any way we can make it so that the snippets we have are actually tested ? Could we pull them out of a unit test somewhere? or a test suite that is specifically samples and we load based on an ID of sorts?

### LINKED ISSUE #623: Transaction Documentation
Related: #621 

Should we privatize `transaction#begin` and `transaction#commit`, removing them from the public docs and their usage in other Transaction examples?

A user goes through their Dataset instance to get a Transaction:

```js
dataset.runInTransaction(transaction, done) {});
```

This calls `transaction.begin` immediately, and the `done` function returned to their callback is actually `transaction.commit`.

Having these methods appear in the docs is probably quite confusing, as the user shouldn't need (or be able to) begin a transaction that already began, or commit the transaction before calling `done` (commit) again.

### PR DESCRIPTION
**Pull request by ryanseys**
Fixes #598
Fixes #622 
Fixes #623 

I really wanted to build this because it seemed like an interesting problem and this builds on our already-epic docs to make them more epic.

This adds an additional test which will run with `npm test` like normal, though I would suggest we consider changing `npm test` to also generate docs, then run the tests to ensure the test remains up to date with the docs generated.

The idea is that we read in `docs/**/*.json` files and iterate through every appropriate `@example` tag, extract the code string, and run it inside a [node.js vm sandbox](https://nodejs.org/api/vm.html), ensuring that it doesn't throw an error along the way. This means that it will check for general syntax errors, as well as calling methods with the wrong number of parameters or missing parameters or just plain incorrect parameters.

A lot of this PR is cleaning up the docs so the test passes! Yeah, it caught a wide range of issues! :+1: 

One thing to note is that it has no way of running anything asynchronously or waiting in anyway for any async function to be called in the VM because we'd have to then inject our own code into the existing examples and that would be quite difficult to guess.

For example `Dataset#runInTransaction` (which you'll see a lot of in this PR) actually is an async function and so it makes it impossible to test anything that is included in the callback of that function, and unfortunately the only way we recommend they create a transaction object is by running `dataset.runInTransaction()` so I'm not sure what we should do to mitigate this issue and test our transaction object. I'll comment on the PR where this issue takes place.

### CODE REVIEW

File: lib/datastore/request.js [hunk 1]
```diff
@@ -258,13 +258,13 @@ DatastoreRequest.prototype.insert = function(entities, callback) {
  *   key: key,
  *   data: {
  *     name: 'DonutShack', // strings
- *     rating: datastore.int(8), // ints
- *     worth: datastore.double(123456.78), // doubles
+ *     rating: gcloud.datastore.int(8), // ints
+ *     worth: gcloud.datastore.double(123456.78), // doubles
  *     numDonutsServed: 45, // detect number type (int or double)
- *     founded: new Date('Tue May 12 2015 15:30:00 GMT-0400 (EDT)'), // dates
+ *     // founded: new Date('Tue May 12 2015 15:30:00 GMT-0400 (EDT)'), // dates
```
Comment (ryanseys) [inline lib/datastore/request.js]: Oops these were commented out because I have no idea why it throws when given a date object. I'll need to investigate this.

File: lib/datastore/transaction.js [hunk 2]
```diff
@@ -85,16 +85,18 @@ nodeutil.inherits(Transaction, DatastoreRequest);
  *     a transaction.
  *
  * @example
- * transaction.begin(function(err) {
- *   // Perform Datastore operations as usual.
- *   transaction.get(dataset.key(['Company', 123]), function(err, entity) {
- *     // Commit the transaction.
- *     transaction.finalize(function(err, apiResponse) {});
+ * dataset.runInTransaction(function(transaction, done) {
```
Comment (ryanseys) [inline lib/datastore/transaction.js]: All of the examples in this file have been wrapped with `dataset.runInTransaction()` however the code inside the callback is not tested because the function is async so for example in this one, `transaction.begin`, `transaction.get` and `transaction.finalize` are never tested (run), except for syntax when the code block is compiled (so it's not a total loss).

File: test/common/docs.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+/*global describe, it */
+
+'use strict';
+
+var vm = require('vm');
+var assert = require('assert');
+var gcloud = require('../../');
+var glob = require('glob');
+var mitm = require('mitm');
```
Comment (ryanseys) [inline test/common/docs.js]: Added `glob` and `mitm` dev deps.

File: test/common/docs.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    throw new Error('\n' + code + '\n\n' + err.message);
+  }
+}
+
+describe('docs', function() {
+  var MITM;
+  var FILES;
+
+  before(function(done) {
+    MITM = mitm();
```
Comment (ryanseys) [inline test/common/docs.js]: Turns off the network so the VM doesn't actually make real requests to google when running the examples. Perhaps I should add a comment here for that?

File: test/common/docs.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    glob('docs/json/**/*.json', {}, function(err, files) {
+      assert.ifError(err);
+      FILES = files;
+      done();
+    });
+  });
+
+  after(function() {
+    MITM.disable();
```
Comment (ryanseys) [inline test/common/docs.js]: Reenable the network.

File: lib/bigquery/job.js [hunk 6]
```diff
@@ -117,12 +118,13 @@ Job.prototype.getMetadata = function(callback) {
...
  * //-
  * var through2 = require('through2');
+ * var fs = require('fs');
  *
  * job.getQueryResults(options)
  *   .pipe(through2.obj(function (row, enc, next) {
  *     this.push(JSON.stringify(row) + '\n');
  *   }))
- *   .pipe(fs.createWriteStream('./backup.json'));
+ *   .pipe(fs.createWriteStream('./test/testdata/testfile.json'));
```
Comment (ryanseys) [inline lib/bigquery/job.js]: These actually write to a file (mind you they write nothing)... so having the file exist in our repo helps mitigate it creating additional files. Is this ok?

File: scripts/build.sh [hunk 7]
```diff
@@ -17,6 +17,7 @@
 set -ev
 
 npm run lint
+npm run docs
```
Comment (ryanseys) [inline scripts/build.sh]: Added this so travis will build docs on normal PRs so it will catch docs issues before they make their way into the repo.

**Comment by jgeewax** [comment]
This is awesome. I totally get it that we can't inspect inside callbacks... maybe there's some magic we can figure out later, but for now this is a huge step forward. :+1:

**Comment by stephenplusplus** [comment]
This is amazing. This is the first time I've seen anything like this in OSS/ever. @jgeewax is there some kind of Google Dev Blog we can write about this on after it's merged?

File: lib/datastore/transaction.js [hunk 8]
```diff
@@ -85,16 +85,18 @@ nodeutil.inherits(Transaction, DatastoreRequest);
...
+ *   transaction.begin(function(err) {
```
Comment (stephenplusplus) [inline lib/datastore/transaction.js]: Remove `transaction.begin` from these examples. Begin is called automatically when `dataset.runInTransaction` is called.

Comment (stephenplusplus) [inline lib/datastore/transaction.js]: I'm getting mixed up if this PR is trying to solve #623. It looks like you wanted to because of wrapping things in `dataset.runInTransaction` and some doc wording changes. If you don't want to solve it in this PR, that's fine, you can leave as is and we can resolve after this merges.

**Comment by ryanseys** [comment]
Adding don't merge so I can continue working on this today to investigate mitigating some of the issues I've made apparent here with the current version.

**Comment by ryanseys** [comment]
Whoa! This docs test found a real live bug! Our `transaction#rollback` didn't attach the transaction id like it should have. I added a unit test to catch it :)

File: lib/datastore/transaction.js [hunk 9]
```diff
@@ -84,16 +89,11 @@ nodeutil.inherits(Transaction, DatastoreRequest);
  * @param {function} callback - The function to execute within the context of
  *     a transaction.
  *
+ * @private
```
Comment (ryanseys) [inline lib/datastore/transaction.js]: Privatize `transaction#begin` without changing the name yet. Should I go ahead and change it to `begin_` to follow convention?

File: lib/datastore/transaction.js [hunk 10]
```diff
@@ -143,10 +141,10 @@ Transaction.prototype.rollback = function(callback) {
  * Commit the remote transaction and finalize the current transaction instance.
  *
  * @param {function} callback - The callback function.
+ * @private
```
Comment (ryanseys) [inline lib/datastore/transaction.js]: Privatize `transaction#commit` without changing the name yet. Should I go ahead and change it to `commit_` to follow convention?

File: test/common/docs.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  it('should run public docs examples without errors', function() {
+    for (var f in FILES) {
+      var fileDocBlocks = require('../../' + FILES[f]);
+      for (var i in fileDocBlocks) {
+        var doc = fileDocBlocks[i];
+
+        // We don't want to test developer docs or private documentation that
+        // are hidden because they use the modules in a less-recommended way
+        // e.g. use class object directly or call obj.makeReq() directly.
+        if (!doc.isPrivate) {
```
Comment (ryanseys) [inline test/common/docs.js]: I am working to remove this restriction so all the docs are tested! :) The more, the better! Death to bitrot!

File: lib/datastore/transaction.js (hunk 9 above)
Comment (stephenplusplus) [inline lib/datastore/transaction.js]: Yeah :+1:

File: lib/datastore/transaction.js (hunk 10 above)
Comment (stephenplusplus) [inline lib/datastore/transaction.js]: Yeah. Maybe our docs with `runInTransaction` should show `done` named as `commit`. Or just explain calling done will perform the commit.

```js
dataset.runInTransaction(function(transaction, commit) {
  // ...
```

File: test/common/docs.js (hunk 11 above)
Comment (stephenplusplus) [inline test/common/docs.js]: Yeah!

File: lib/datastore/transaction.js [hunk 12]
```diff
@@ -42,6 +42,11 @@ var extend = require('extend');
  *     connection to Google Cloud Datastore.
  * @param {string} projectId - Dataset ID. This is your project ID from the
  *     Google Developers Console.
+ *
+ * @example
+ * var Transaction = require('gcloud/lib/datastore/transaction');
```
Comment (ryanseys) [inline lib/datastore/transaction.js]: This is added to the developer documentation, which is hidden from generated docs (as far as I know), however because the VM runs this with the other examples in the file, we now have a proper `transaction` variable that can be used for all the examples that follow.

**Comment by stephenplusplus** [comment]
> Whoa! This docs test found a real live bug! Our transaction#rollback didn't attach the transaction id like it should have. I added a unit test to catch it :)

Nice!

File: lib/datastore/transaction.js (hunk 10 above)
Comment (ryanseys) [inline lib/datastore/transaction.js]: :+1:

File: lib/datastore/transaction.js (hunk 12 above)
Comment (stephenplusplus) [inline lib/datastore/transaction.js]: Yeah, dev docs are hidden because of the `/*!` header. Can you add a comment / line in the dev doc block about why that's there?

Comment (ryanseys) [inline lib/datastore/transaction.js]: Will do

File: lib/datastore/transaction.js (hunk 10 above)
Comment (ryanseys) [inline lib/datastore/transaction.js]: Why do we provide both? Why not just tell them to run `transaction.commit()` ?

Comment (stephenplusplus) [inline lib/datastore/transaction.js]: I think we did it to be Node convention-friendly. Async functions usually gets a function to execute when you're done.

Comment (ryanseys) [inline lib/datastore/transaction.js]: Ah, ok. This makes sense.

**Comment by ryanseys** [comment]
PTAL. Should test all our examples, including ignored and private ones. 
Also privatized `Transaction#begin_` and `Transaction#commit_`.

File: test/common/docs.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    });
+  });
+
+  after(function() {
+    // Re-enable the network.
+    MITM.disable();
+  });
+
+  it('should run docs examples without errors', function() {
+    for (var f in FILES) {
```
Comment (stephenplusplus) [inline test/common/docs.js]: `FILES` is an array isn't it? You can do FILES.forEach.

File: test/common/docs.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+      var fileContents = fs.readFileSync(FILES[f], {
+        encoding: 'utf8'
+      });
+
+      try {
+        fileDocBlocks = JSON.parse(fileContents);
+      } catch(e) {
+        throw new Error('Failed to parse one of the doc files (' + e.message +
+            ')\n\nFilename: ' + FILES[f] + '\n\nFile contents: \'' +
+            fileContents + '\'');
```
Comment (stephenplusplus) [inline test/common/docs.js]: Can you make this message an array, maybe joined by '\n\n', so it's a little easier to look at?

Comment (stephenplusplus) [inline test/common/docs.js]: Maybe `"` instead of `'` to avoid the escaping will help as well.

File: test/common/docs.js [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+      }
+
+      for (var i in fileDocBlocks) {
```
Comment (stephenplusplus) [inline test/common/docs.js]: `fileDocBlocks.forEach`

File: test/common/docs.js [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+        var tags = fileDocBlocks[i].tags;
+        for (var j in tags) {
```
Comment (stephenplusplus) [inline test/common/docs.js]: Not sure if `tags` is an array, but `tags.forEach` if it is.

File: test/common/docs.js (hunk 13 above)
Comment (ryanseys) [inline test/common/docs.js]: How am I just learning about this now?

giphy

File: test/common/docs.js [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+describe('documentation', function() {
+  var MITM;
+  var FILES;
+
+  before(function(done) {
+    // Turn off the network so that API calls aren't actually made.
+    MITM = mitm();
+    glob('docs/json/**/*.json', {}, function(err, files) {
```
Comment (stephenplusplus) [inline test/common/docs.js]: The empty `{}` isn't necessary.

File: lib/datastore/request.js [hunk 18]
```diff
@@ -465,17 +460,13 @@ DatastoreRequest.prototype.delete = function(keys, callback) {
  *     stream instance is returned.
  *
  * @example
- * //-
- * // Where you see `transaction`, assume this is the context that's relevant to
- * // your use, whether that be a Dataset or a Transaction object.
- * //-
```
Comment (stephenplusplus) [inline lib/datastore/request.js]: Can you run me through why some of the text of these are changed and others are removed?

File: lib/datastore/request.js [hunk 19]
```diff
@@ -465,17 +460,13 @@ DatastoreRequest.prototype.delete = function(keys, callback) {
...
  *
  * // Retrieve 5 companies.
- * transaction.runQuery(query, function(err, entities, endCursor, apiResponse) {
+ * dataset.runQuery(query, function(err, entities, endCursor, apiResponse) {
```
Comment (stephenplusplus) [inline lib/datastore/request.js]: I don't see where `query` is defined here. Should we make a fake one? Also, shouldn't the new test suite have thrown on this?

Comment (jgeewax) [inline lib/datastore/request.js]: +1 - seems weird.. Are we properly scoping variables between examples? Maybe it's picking up a previously defined `query` variable?

**Comment by stephenplusplus** [comment]
This isn't a problem in this PR, but I just discovered something I must have messed up before:

https://github.com/ryanseys/gcloud-node/blob/verify-docs/lib/datastore/transaction.js#L266

Those examples shouldn't show `delete` taking a callback. Feel free to fix while you're in there, or I can always get to it after.

File: lib/datastore/request.js (hunk 19 above)
Comment (ryanseys) [inline lib/datastore/request.js]: You're right JJ, this is likely because the sandbox is shared between files meaning that any values defined in one are accessible in another. I will update the test to use a new sandbox for each file as long as it's relatively easy to migrate our existing code over.

File: lib/datastore/request.js (hunk 18 above)
Comment (ryanseys) [inline lib/datastore/request.js]: I think this is a result of originally not having a transaction object. I will fix this.

Comment (ryanseys) [inline lib/datastore/request.js]: Why are we using `transaction` in these examples anyway instead of `dataset`?

File: test/common/docs.js [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+        fileDocBlocks = JSON.parse(fileContents);
+      } catch(e) {
+        throw new Error([
+          'Failed to parse one of the doc files (' + e.message + ')',
+          'Filename: ' + filename,
+          'File contents: "' + fileContents + '"'
+        ].join('\n'));
+      }
+
+      var sandbox = {
```
Comment (ryanseys) [inline test/common/docs.js]: Using a fresh sandbox for each file now.

File: lib/datastore/transaction.js [hunk 21]
```diff
@@ -260,13 +264,13 @@ Transaction.prototype.commit = function(callback) {
  *
  * @example
  * // Delete a single entity.
- * transaction.delete(dataset.key(['Company', 123]), function(err) {});
+ * transaction.delete(dataset.key(['Company', 123]));
```
Comment (ryanseys) [inline lib/datastore/transaction.js]: @stephenplusplus Removed callbacks.

**Comment by ryanseys** [comment]
PTAL.

File: test/common/docs.js (hunk 20 above)
Comment (stephenplusplus) [inline test/common/docs.js]: :+1:

File: lib/datastore/dataset.js [hunk 22]
```diff
@@ -208,7 +210,10 @@ Dataset.prototype.createQuery = function(namespace, kind) {
 
 /**
  * Run a function in the context of a new transaction. Transactions allow you to
- * perform multiple operations, committing your changes atomically.
+ * perform multiple operations, committing your changes atomically. When you are
+ * finished making your changes within the transaction, run the done() function
+ * provided in the callback function to commit your changes. See an example
+ * below for more information.
```
Comment (stephenplusplus) [inline lib/datastore/dataset.js]: :+1:

File: lib/datastore/entity.js [hunk 23]
```diff
@@ -411,7 +411,7 @@ function valueToProperty(v) {
     p.blob_value = v;
     return p;
   }
-  if (v instanceof Array) {
+  if (Array.isArray(v)) {
```
Comment (stephenplusplus) [inline lib/datastore/entity.js]: :+1: :+1:

File: lib/datastore/request.js [hunk 24]
```diff
@@ -664,6 +680,10 @@ DatastoreRequest.prototype.makeReq_ = function(method, body, callback) {
     }
   }
 
+  if (method === 'rollback') {
+    body.transaction = this.id;
```
Comment (stephenplusplus) [inline lib/datastore/request.js]: :+1: :+1: :+1:

File: lib/datastore/transaction.js [hunk 25]
```diff
@@ -119,12 +123,10 @@ Transaction.prototype.begin = function(callback) {
  * @param {function} callback - The callback function.
  *
  * @example
- * transaction.begin(function(err) {
- *   transaction.rollback(function(err) {
- *     if (err) {
- *       // Transaction could not be rolled back.
- *     }
- *   });
+ * transaction.rollback(function(err) {
```
Comment (stephenplusplus) [inline lib/datastore/transaction.js]: This is actually a bit confusing, because a rollback can only be done on a transaction that has been committed (I think?), but that's a problem:

```js
dataset.runInTransaction(function(transaction, _commit) {
  
}, function(err) {
  // no `transaction` here to call `rollback` on.
})
```

Also, is there a use case for a user wanting to rollback a transaction that they didn't just create? 

Sorry all this transaction stuff is getting lumped into this PR.

File: lib/datastore/transaction.js (hunk 21 above)
Comment (stephenplusplus) [inline lib/datastore/transaction.js]: Thanks!

File: lib/storage/bucket.js [hunk 26]
```diff
@@ -65,7 +65,7 @@ var RESUMABLE_THRESHOLD = 5000000;
  * @constructor
  * @alias module:storage/bucket
  *
- * @throws if a bucket name isn't provided.
+ * @throws {Error} if a bucket name isn't provided.
```
Comment (stephenplusplus) [inline lib/storage/bucket.js]: Debugging crown goes to @ryanseys :crown:

File: lib/storage/bucket.js [hunk 27]
```diff
@@ -218,14 +219,16 @@ function Bucket(storage, name) {
  * @param {function=} callback - The callback function.
  *
  * @example
- * var 2013logs = bucket.file('2013-logs.txt');
- * var 2014logs = bucket.file('2014-logs.txt');
```
Comment (stephenplusplus) [inline lib/storage/bucket.js]: I'll learn JavaScript one of these days.

File: package.json [hunk 28]
```diff
@@ -76,7 +78,7 @@
   "scripts": {
     "docs": "./scripts/docs.sh",
     "lint": "jshint lib/ system-test/ test/",
-    "test": "mocha test/*",
+    "test": "npm run docs && mocha test/*",
```
Comment (stephenplusplus) [inline package.json]: Now that's how tests should be done!

File: lib/datastore/transaction.js (hunk 25 above)
Comment (ryanseys) [inline lib/datastore/transaction.js]: Can you open a separate issue for this?

File: lib/storage/bucket.js (hunk 26 above)
Comment (ryanseys) [inline lib/storage/bucket.js]: If you call *deleting code until it passes then seeing what you deleted to find the issue* debugging.

File: lib/datastore/transaction.js (hunk 25 above)
Comment (ryanseys) [inline lib/datastore/transaction.js]: Just cuz I don't know this issue on hand and like you said, doesn't really belong in this PR.

File: lib/storage/bucket.js (hunk 26 above)
Comment (stephenplusplus) [inline lib/storage/bucket.js]: That's *exactly* what I call debugging.

File: lib/datastore/transaction.js (hunk 25 above)
Comment (stephenplusplus) [inline lib/datastore/transaction.js]: Of course :) #633

File: lib/storage/bucket.js (hunk 27 above)
Comment (ryanseys) [inline lib/storage/bucket.js]: While we're requesting features for the JS tokenizer, I'd like to request supporting hyphens.

**Comment by stephenplusplus** [comment]
This is awesome. You made history.

**Comment by ryanseys** [comment]
Yay! This thing is awesome! :) I'm excited to see it break our builds with no remorse.
